### PR TITLE
issue #55 - Draw callback error

### PR DESCRIPTION
**Pull request by daymosik**
https://github.com/elm-lang/virtual-dom/issues/55

### CODE REVIEW

**Comment by witoldsz** [comment]
As far as I can see, it solves the problem with IE9 and old Chrome browsers (tested on version 19) which is similar to Android 4.1 stock browsers.

File: src/Native/VirtualDom.js [hunk 1]
```diff
@@ -1535,10 +1535,14 @@ function normalRenderer(parentNode, view)
 
 // STEPPER
 
-var rAF =
-	typeof requestAnimationFrame !== 'undefined'
-		? requestAnimationFrame
-		: function(callback) { callback(); };
+var rAF = window.requestAnimFrame = (function(){
```
Comment (witoldsz) [inline src/Native/VirtualDom.js]: The `= window.requestAnimFrame` is not required, is it?

Comment (witoldsz) [inline src/Native/VirtualDom.js]: the *self-executing anonymous function*: `(function() {…})()` could be removed, I guess…

Comment (daymosik) [inline src/Native/VirtualDom.js]: i guess it's not necessary.

Comment (daymosik) [inline src/Native/VirtualDom.js]: but this is required for IE9, Android Browsers etc.

**Comment by daymosik** [comment]
cleaner version. i looked and it worked the same.

**Comment by evancz** [comment]
Thanks @rtfeldman for bringing this to my attention. Happy to do a patch to get this out for folks.

I have one question though. Why does it work? 

Surely it is related to this part:

```javascript
// old
function(callback) { callback(); }

// new
function(callback) {
    window.setTimeout(callback, 1000 / 60);
};
```

Why is a `setTimeout` needed? I took it out because it does not really give you the same benefits as `requestAnimationFrame`. Instead it just introduces a delay of 16 milliseconds to every single thing you do.

**Comment by evancz** [comment]
Oh, I thought of another important question. With this change, are you able to run the following in the REPL:

```elm
> import Html
> Html.text "hello"
```

Using things like `window` and `document` directly can cause issues when the REPL uses node to evaluate things. We should be careful not to fix this at the cost of another bug.

**Comment by witoldsz** [comment]
@evancz Maybe this can help: https://github.com/chrisdickinson/raf
It's a polyfill which makes Elm 0.18 work in old browsers like the ones in Android 4.1–4.3. We are successfully using this workaround on a production now. I don't know where did @daymosik find his solution, but the `setTimeout` is there too…

**Comment by evancz** [comment]
No. I understand *what* was changed. I'm asking:

  - Why is `setTimeout` needed?
  - Does this PR work with `elm-repl`?

Please keep further discussion focused on these technical questions.

**Comment by Mouvedia** [comment]
> [For example, if the browser is attempting to achieve a 60Hz refresh rate, then these steps are only necessary every 60th of a second (about **16.7**ms).](https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-processing-model:top-level-browsing-context)

**Comment by rtfeldman** [comment]
@Mouvedia that appears to be a document about how browser authors should implement event loops, nothing to do with client-side JavaScript...so I'm not sure it's pertinent here. :)

Separately, "these steps are only necessary" suggests the worst-case expected downside here would be wasted CPU cycles, not a runtime exception.

**Comment by evancz** [comment]
@Mouvedia, that's not relevant information.

Sorry to be curt with folks. Changing how this works is a big deal, so I'm frustrated by the empty description on this PR and general lack of relevant technical details. Maybe [the checklist](https://github.com/process-bot/contribution-checklist/blob/master/pulls.md#checklist) provided by @process-bot needs to be refined.

I've now gone and found these details, so here they are for everyone else.


## 1 `setTimeout`

Rendering is managed by [`updateIfNeeded`](https://github.com/elm-lang/virtual-dom/blob/master/src/Native/VirtualDom.js#L1549). It runs a state machine that helps us get really smooth animations when using `requestAnimationFrame`. Whenever it does work, it asks for the next frame as well. That may not be necessary, but if you do not do it immediately, it is possible to "drop frames". If you call `requestAnimationFrame` too late, it actually waits an extra round. This state machine has been shown to get really nice results!

The problem is with preemptively asking for the next frame [here](https://github.com/elm-lang/virtual-dom/blob/master/src/Native/VirtualDom.js#L1560). Assuming I'm reading it correctly, this would go into an infinite loop if `raf = function(f) { f(); }`.

So that is *why* a `setTimeout` is needed in our case. I don't have devices that'd let me test whether 0ms or 16ms is a better choice. If someone can test (maybe @witoldsz?) that'd be great! Would much prefer to give the more efficient of the two!

## 2 `elm-repl`

Based on [this file](https://github.com/elm-lang/elm-repl/blob/master/src/Eval/Code.hs) in the REPL, I'm pretty sure these changes will not work with `elm-repl` as is. I suspect we will see an error like: `window` is not defined.

If you can change these lines in your `elm-stuff/packages/elm-lang/virtual-dom/2.0.3/src/Native/VirtualDom.js` file, you can see if they work locally. I will try that, but it'd be great to get confirmation of my results.

**Comment by evancz** [comment]
I've suggested an alternate fix that avoids the `elm-repl` issues at https://github.com/elm-lang/virtual-dom/pull/78

Can folks confirm that it works?

**Comment by Mouvedia** [comment]
@rtfeldman just wanted to illustrate with an official source the seemingly arbitrary `1000 / 60`. Nowadays everything seems to be synced on that value. It probably has no consequence in your case though. But it's sometime important to reduce the number of redraws. Maybe it was immediately obvious to you that it referred to 60Hz, but it wasn't to me.

**Comment by daymosik** [comment]
@evancz i've made a fix in my PR and i think we should use the browsers prefixed requestAnimationFrame also, in order to support some older browsers people may be still using.

File: src/Native/VirtualDom.js [hunk 2]
```diff
@@ -1534,11 +1534,11 @@ function normalRenderer(parentNode, view)
...
 // STEPPER
-
 var rAF =
-	typeof requestAnimationFrame !== 'undefined'
-		? requestAnimationFrame
-		: function(callback) { callback(); };
+  typeof requestAnimationFrame !== 'undefined' ? requestAnimationFrame :
+  typeof webkitRequestAnimationFrame !== 'undefined' ? webkitRequestAnimationFrame :
+  typeof mozRequestAnimationFrame !== 'undefined' ? mozRequestAnimationFrame :
+  function(callback) { setTimeout(callback, 1000 / 60); };
```
Comment (Mouvedia) [inline src/Native/VirtualDom.js]: Opera and Microsoft vendor prefixes are missing.

```
window.requestAnimationFrame =	requestAnimationFrame		||
								mozRequestAnimationFrame		||
								webkitRequestAnimationFrame	||
								oRequestAnimationFrame		||
								msRequestAnimationFrame;
```

Comment (daymosik) [inline src/Native/VirtualDom.js]: i would not really rely on microsoft's experimental implementation...

Comment (Mouvedia) [inline src/Native/VirtualDom.js]: @daymosik You would be wrong; the one you should be wary of is the Mozilla one.
cf https://bugzilla.mozilla.org/show_bug.cgi?id=630127

**Comment by evancz** [comment]
This was fixed in #78

**Pull request closed without merging**